### PR TITLE
ARROW-350: Added Kerberos to HDFS client

### PR DESCRIPTION
**Pull request by chrisaycock**

### CODE REVIEW

File: cpp/src/arrow/io/hdfs.h [hunk 1]
```diff
@@ -60,19 +60,16 @@ struct HdfsConnectionConfig {
   std::string host;
   int port;
   std::string user;
-
-  // TODO: Kerberos, etc.
+  std::string kerb;
```
Comment (wesm) [inline cpp/src/arrow/io/hdfs.h]: Can you give this a more descriptive name, like `kerberos_ticket_path`

File: python/pyarrow/io.pyx [hunk 2]
```diff
@@ -313,14 +313,21 @@ cdef class HdfsClient:
...
 
         Parameters
         ----------
-        host :
-        port :
-        user :
+        host : NameNode. Set to "default" for fs.defaultFS from core-site.xml.
+        port : NameNode's port. Set to 0 for default or logical (HA) nodes.
+        user : Username when connecting to HDFS; None implies login user.
+        kerb : Path to Kerberos ticket cache.
```
Comment (wesm) [inline python/pyarrow/io.pyx]: Same comment as above

File: python/pyarrow/io.pyx [hunk 3]
```diff
@@ -335,9 +342,13 @@ cdef class HdfsClient:
             HdfsClient out = HdfsClient()
             HdfsConnectionConfig conf
 
-        conf.host = tobytes(host)
+        if host is not None:
+            conf.host = tobytes(host)
```
Comment (wesm) [inline python/pyarrow/io.pyx]: Is `None` a valid option?

**Review by wesm — CHANGES_REQUESTED** [review]

**Comment by chrisaycock** [comment]
A `None` host is valid from the standpoint of libhdfs. Here's the doc from `hdfs.h`:

> If NULL is given, a LocalFileSystem will be created.

I'm not sure why they chose to specify the default host with `"default"`, but I wanted to keep my parameter values consistent with the C++.

I'll rename the `kerb` parameter.

**Comment by wesm** [comment]
+1

**Pull request closed without merging**